src/librvnpal/src/posix/mapping.c

```diff
@@ -0,0 +1,249 @@
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <sys/mman.h>
+#include <string.h>
+#include <assert.h>
+
+#include "rvn.h"
+#include "status_codes.h"
+#include "internal_posix.h"
+
+struct map_file_handle
+{
+    int fd;
+    const char* path;
+    int flags;
+};
+
+PRIVATE void free_map_file_handle(struct map_file_handle *handle)
+{
+    if (handle->path != NULL)
+    {
+        free((void *)(handle->path));
+        (handle)->path = NULL;
+    }
+
+    free((void *)handle);
+}
+
+EXPORT int32_t
+rvn_create_and_mmap64_file(
+    const char *path,
+    int64_t initial_file_size,
+    int32_t flags,
+    void **handle,
+    void **base_addr,
+    int64_t *actual_file_size,
+    int32_t *detailed_error_code)
+{
+    int32_t rc = SUCCESS;
+
+    assert(initial_file_size > 0);
+
+    _ensure_path_exists(path);
+      
+    struct map_file_handle* mfh = calloc(1, sizeof(struct map_file_handle));
+    *handle = mfh;
+    if(mfh == NULL)
+    {
+        rc = FAIL_CALLOC;
+        goto error_cleanup;
+    }
+    mfh->path = strdup(path);
+    if (mfh->path == NULL)
+    {
+        rc = FAIL_CALLOC;
+        goto error_cleanup;
+    }
+    mfh->flags = flags;
+    mfh->fd = open(path, O_RDWR | O_CREAT, S_IWUSR | S_IRUSR);
+    if (mfh->fd == -1)
+    {
+        rc = FAIL_OPEN_FILE;
+        goto error_cleanup;
+    }
+
+    int64_t sz = lseek(mfh->fd, 0, SEEK_END);
+    if (sz == -1)
+    {
+        rc = FAIL_SEEK_FILE;
+        goto error_cleanup;
+    }
+
+    int64_t sys_page_size = sysconf(_SC_PAGE_SIZE);
+    if (sys_page_size == -1)
+    {
+        rc = FAIL_SYSCONF;
+        goto error_cleanup;
+    }
+
+    int32_t allocation_granularity = ALLOCATION_GRANULARITY;
+    if (initial_file_size < allocation_granularity)
+        initial_file_size = allocation_granularity;
+
+    if (sz <= initial_file_size || sz % sys_page_size != 0)
```

Comment:
Here you are _always_ syncing the directory on opening a file, why?
We only need to do that if the file size changed.

Reply:
Or the file was created